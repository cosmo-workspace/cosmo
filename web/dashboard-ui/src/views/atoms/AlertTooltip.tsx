import {
  darken, lighten, styled, Tooltip, tooltipClasses, TooltipProps
} from "@mui/material";
import React from "react";

export const AlertTooltip = styled(({ className, ...props }: TooltipProps) => (
  <Tooltip {...props} arrow classes={{ popper: className }} />
))(({ theme }) => ({
  [`& .${tooltipClasses.arrow}`]: {
    fontSize: '1rem',
    color: (theme.palette.mode === 'light' ? lighten : darken)(theme.palette['info'].light, 0.9),
  },
  [`& .${tooltipClasses.tooltip}`]: {
    borderRadius: "4px",
    boxShadow: theme.shadows[5],
    backgroundColor: (theme.palette.mode === 'light' ? lighten : darken)(theme.palette['info'].light, 0.9),
  },
}));
