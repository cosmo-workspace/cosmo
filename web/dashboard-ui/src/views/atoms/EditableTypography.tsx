import { Check, Close, Edit } from "@mui/icons-material";
import {
  IconButton,
  InputBase,
  InputBaseProps,
  Stack,
  Tooltip,
} from "@mui/material";
import React, { useState } from "react";

export type EditableTypographyProps = InputBaseProps & {
  children: string;
  onSave: (inputData: string) => void;
  showAlways?: boolean;
};

export const EditableTypography: React.FC<EditableTypographyProps> = ({
  children,
  onSave,
  showAlways,
  ...props
}) => {
  const [showEditIcon, setShowEditIcon] = useState(showAlways);
  const [editting, setEditing] = useState(false);
  const [inputData, setInputData] = useState<string>(children);

  const editIcon = (
    <Tooltip title="Edit" placement="top">
      <IconButton
        disableRipple
        size="small"
        type="button"
        sx={{ ml: 1 }}
        onClick={() => {
          setEditing(true);
        }}
      >
        <Edit />
      </IconButton>
    </Tooltip>
  );
  const editingIcons = (
    <Stack direction="row">
      <Tooltip title="Save" placement="top">
        <IconButton
          disableRipple
          size="small"
          type="button"
          sx={{ ml: 1 }}
          onClick={() => {
            onSave(inputData);
            setEditing(false);
          }}
        >
          <Check />
        </IconButton>
      </Tooltip>
      <Tooltip title="Cancel" placement="top">
        <IconButton
          disableRipple
          size="small"
          type="button"
          sx={{ ml: 1 }}
          onClick={() => {
            setInputData(children);
            setEditing(false);
          }}
        >
          <Close />
        </IconButton>
      </Tooltip>
    </Stack>
  );

  return (
    <InputBase
      size="small"
      fullWidth
      placeholder="Name"
      endAdornment={
        editting ? editingIcons : showEditIcon ? editIcon : undefined
      }
      defaultValue={children}
      readOnly={!editting}
      autoFocus={editting}
      onMouseOver={() => {
        setShowEditIcon(showAlways || true);
      }}
      onMouseLeave={() => {
        setShowEditIcon(showAlways || false);
      }}
      onBlur={(e) => {
        setInputData(e.currentTarget.value);
      }}
      sx={{ borderBottom: editting ? 1 : 0 }}
      {...props}
    />
  );
};
