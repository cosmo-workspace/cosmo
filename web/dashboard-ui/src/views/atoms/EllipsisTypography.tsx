import {
  Tooltip,
  TooltipProps,
  Typography,
  TypographyProps,
} from "@mui/material";
import React from "react";

export type EllipsisTypographyProps = Omit<TypographyProps, "children"> & {
  children?: string;
  placement?: TooltipProps["placement"];
};

export const EllipsisTypography: React.FC<EllipsisTypographyProps> = (
  props
) => {
  const [isOverflow, setIsOverflow] = React.useState(false);
  const paragraph = React.useRef<HTMLSpanElement>(null);

  React.useEffect(() => {
    const pElement = paragraph.current;
    if (pElement) {
      setIsOverflow(Boolean(pElement.offsetWidth < pElement.scrollWidth));
    }
  }, [paragraph]);

  const title = props.children || "";

  // return isOverflow ? (<Tooltip title={title} placement={placement} > {typoglaphy} </Tooltip >) : typoglaphy
  return isOverflow ? (
    <Tooltip title={title} placement={props.placement}>
      <Typography
        ref={paragraph}
        variant="caption"
        display="block"
        sx={{
          textOverflow: "ellipsis",
          whiteSpace: "nowrap",
          overflow: "hidden",
        }}
        {...props}
      >
        {props.children}
      </Typography>
    </Tooltip>
  ) : (
    <Typography
      ref={paragraph}
      variant="caption"
      display="block"
      sx={{
        textOverflow: "ellipsis",
        whiteSpace: "nowrap",
        overflow: "hidden",
      }}
      {...props}
    >
      {props.children}
    </Typography>
  );
};
