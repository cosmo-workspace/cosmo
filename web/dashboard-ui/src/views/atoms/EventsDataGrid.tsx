import { Info, Warning } from "@mui/icons-material";
import { Chip, SxProps } from "@mui/material";
import {
  DataGrid,
  DataGridProps,
  GridColDef,
  gridClasses,
} from "@mui/x-data-grid";
import React from "react";
import { Event } from "../../proto/gen/dashboard/v1alpha1/event_pb";
import { EventDetailDialogContext } from "../organisms/EventDetailDialog";

export type EventsDataGridProp = {
  events: Event[];
  maxHeight?: number;
  sx?: SxProps;
  dataGridProps?: Omit<DataGridProps, "columns" | "sx">;
  clock: Date;
};

export const EventsDataGrid: React.FC<EventsDataGridProp> = ({
  events,
  maxHeight,
  sx,
  dataGridProps,
  clock,
}) => {
  const eventDetailDialogDispatch = EventDetailDialogContext.useDispatch();

  const columns: GridColDef[] = [
    { field: "type", headerName: "Type", width: 50 },
    {
      field: "eventTime",
      headerName: "LastSeen",
      valueGetter: (value, row: Event) =>
        (row.series?.lastObservedTime || row.eventTime)?.toDate(),
      valueFormatter: (value: Date) =>
        formatTime(clock.getTime() - value.getTime()),
      flex: 0.2,
      minWidth: 80,
    },
    {
      field: "reason",
      headerName: "Reason",
      flex: 0.25,
      renderCell: (params) => (
        <Chip
          icon={
            params.api.getRow(params.id).type == "Normal" ? (
              <Info color="success" />
            ) : (
              <Warning color="warning" />
            )
          }
          label={params.value}
        />
      ),
    },
    { field: "regardingWorkspace", headerName: "Workspace" },
    {
      field: "regarding",
      headerName: "Object",
      valueGetter: (value, row) =>
        row.regarding?.kind + "/" + row.regarding?.name,
      flex: 0.3,
    },
    { field: "reportingController", headerName: "Reporter" },
    {
      field: "series",
      headerName: "Count",
      valueGetter: (value, row) => row.series?.count || 1,
    },
    { field: "note", headerName: "Message", flex: 1 },
  ];

  return (
    <>
      <div style={{ width: "100%", maxHeight: maxHeight, minHeight: 100 }}>
        <DataGrid
          autoHeight={maxHeight === undefined}
          rows={events}
          columns={columns}
          getRowHeight={() => "auto"}
          sx={{
            [`& .${gridClasses.cell}`]: {
              py: 1,
            },
            ...sx,
          }}
          onRowDoubleClick={(params) => {
            eventDetailDialogDispatch(true, { event: params.row });
          }}
          hideFooter
          {...dataGridProps}
        />
      </div>
    </>
  );
};

export function formatTime(milisec: number): string {
  const sec = Math.floor(milisec / 1000) % 60;
  const min = Math.floor(milisec / 1000 / 60) % 60;
  const hours = Math.floor(milisec / 1000 / 60 / 60) % 24;
  if (hours > 0) return `${hours}h${min}m${sec}s`;
  if (min > 0) return `${min}m${sec}s`;
  return `${sec}s`;
}
