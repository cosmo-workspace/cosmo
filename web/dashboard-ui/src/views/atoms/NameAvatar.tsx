import { Avatar, AvatarProps, Typography } from "@mui/material";
import { AccountCircle } from "@mui/icons-material";
import React from "react";

export const NameAvatar: React.VFC<{ name?: string } & AvatarProps> = (props) => {
  return (
    props.name ?
      <Avatar {...props} sx={{ bgcolor: stringToColor(props.name), ...props.sx }}>
        <Typography sx={{ color: (theme) => (theme.palette.mode === 'light' ? 'white' : 'black') }} fontSize='inherit'>
          {props.name.substring(0, 1).toUpperCase()}
        </Typography>
      </Avatar>
      :
      <Avatar {...props} ><AccountCircle /></Avatar>
  )
}

const stringToColor = (str: string) => {
  let hash = 0;
  for (let i = 0; i < str.length; i++) {
    hash = str.charCodeAt(i) + ((hash << 5) - hash);
  }
  let color = '#';
  for (let i = 0; i < 3; i++) {
    let value = (hash >> (i * 8)) & 0xFF;
    color += ('00' + value.toString(16)).substr(-2);
  }
  return color;
}
