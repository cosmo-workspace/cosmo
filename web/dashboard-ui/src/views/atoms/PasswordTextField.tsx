import {
  IconButton,
  InputAdornment, TextField, TextFieldProps
} from "@mui/material";
import { Visibility, VisibilityOff, VpnKey } from "@mui/icons-material";
import React, { useState } from "react";


export const PasswordTextField: React.VFC<TextFieldProps> = ({ type, ...props }) => {
  const [isPassShow, setIsPassShow] = useState(false);

  return (
    <TextField {...props} type={isPassShow ? 'text' : 'password'}
      InputProps={{
        startAdornment: (<InputAdornment position="start"><VpnKey /></InputAdornment>),
        endAdornment: (<InputAdornment position="end">
          <IconButton
            tabIndex={-1}
            onPointerDown={() => { setIsPassShow(true) }}
            onPointerUp={() => { setIsPassShow(false) }}
            onPointerOut={() => { setIsPassShow(false) }}
          >
            {isPassShow ? <Visibility /> : <VisibilityOff />}
          </IconButton>
        </InputAdornment>),
      }}
    />
  );
}
