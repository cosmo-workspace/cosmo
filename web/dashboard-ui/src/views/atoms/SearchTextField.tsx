import { Clear, SearchTwoTone } from "@mui/icons-material";
import {
  IconButton,
  InputAdornment,
  TextField,
  TextFieldProps,
} from "@mui/material";
import React from "react";

export const SearchTextField: React.FC<
  {
    search: string;
    setSearch: (search: string) => void;
  } & TextFieldProps
> = ({ search, setSearch, ...props }) => {
  return (
    <TextField
      InputProps={
        search !== ""
          ? {
              startAdornment: (
                <InputAdornment position="start">
                  <SearchTwoTone />
                </InputAdornment>
              ),
              endAdornment: (
                <InputAdornment position="end">
                  <IconButton
                    size="small"
                    tabIndex={-1}
                    onClick={() => {
                      setSearch("");
                    }}
                  >
                    <Clear />
                  </IconButton>
                </InputAdornment>
              ),
            }
          : {
              startAdornment: (
                <InputAdornment position="start">
                  <SearchTwoTone />
                </InputAdornment>
              ),
            }
      }
      placeholder="Search"
      size="small"
      value={search}
      onChange={(e) => setSearch(e.target.value)}
      sx={{ flexGrow: 0.5 }}
      {...props}
    />
  );
};
