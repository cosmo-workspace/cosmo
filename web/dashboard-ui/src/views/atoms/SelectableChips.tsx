import { Check } from "@mui/icons-material";
import { Chip, ChipProps, ChipTypeMap } from "@mui/material";
import { forwardRef, useState } from "react";
import { useController, UseControllerProps } from "react-hook-form";

type toggleChipProps = { variant: ChipTypeMap['props']['variant'], onClick?: () => void } & ChipProps

export const FormSelectableChip = forwardRef((props: UseControllerProps<any> & ChipProps, ref) => {
    const { field } = useController(props);

    return <SelectableChip onChecked={field.onChange} onBlur={field.onBlur} {...props} />
})

export const SelectableChip: React.FC<{ checked?: boolean, onChecked?: (...event: any[]) => void } & ChipProps> = ({ onChecked, ...props }) => {
    const [mouseovered, setMouseovered] = useState(false);

    const [checked, setChecked] = useState<boolean>(props.defaultChecked || false);

    if (props.checked !== undefined && props.checked !== checked) {
        setChecked(props.checked);
    }

    const toggleChecked = () => { setChecked(!checked); onChecked && onChecked(!checked); }

    const unCheckedChipProps: toggleChipProps = { variant: "outlined", onClick: toggleChecked }
    const checkedChipProps: toggleChipProps = { variant: "filled", onClick: toggleChecked, onDelete: toggleChecked, deleteIcon: <Check /> }
    const checkedChipPropsMouseovered: toggleChipProps = { variant: "filled", onClick: toggleChecked, onDelete: toggleChecked }

    return <Chip {...(checked
        ? (mouseovered ? checkedChipPropsMouseovered : checkedChipProps)
        : unCheckedChipProps)}
        onMouseEnter={() => { setMouseovered(true) }}
        onMouseLeave={() => { setMouseovered(false) }}
        {...props} />
}
