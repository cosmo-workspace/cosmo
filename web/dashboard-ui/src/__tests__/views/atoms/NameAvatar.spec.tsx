import { createTheme, ThemeProvider } from "@mui/material/styles";
import { cleanup, render } from '@testing-library/react';
import React from 'react';
import { afterEach, describe, expect, it } from "vitest";
import { NameAvatar } from '../../../views/atoms/NameAvatar';

afterEach(cleanup);

describe('NameAvatar', () => {

  it('NameAvatar', () => {
    const { asFragment } = render(
      <NameAvatar />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('NameAvatar name', () => {
    const { asFragment } = render(
      <NameAvatar name='cosmo' />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('NameAvatar name dark', () => {
    const myTheme = createTheme({
      palette: { mode: 'dark' },
    });

    const { asFragment } = render(
      <ThemeProvider theme={myTheme}>
        <NameAvatar name='cosmo' />,
      </ThemeProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
