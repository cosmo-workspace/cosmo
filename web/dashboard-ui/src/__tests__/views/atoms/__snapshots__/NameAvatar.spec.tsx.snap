// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NameAvatar > NameAvatar 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 40px;
  height: 40px;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-size: 1.25rem;
  line-height: 1;
  border-radius: 50%;
  overflow: hidden;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #fff;
  background-color: #bdbdbd;
}

.emotion-1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
}

<div
    class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault emotion-0"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-1"
      data-testid="AccountCircleIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 4c1.93 0 3.5 1.57 3.5 3.5S13.93 13 12 13s-3.5-1.57-3.5-3.5S10.07 6 12 6zm0 14c-2.03 0-4.43-.82-6.14-2.88C7.55 15.8 9.68 15 12 15s4.45.8 6.14 2.12C16.43 19.18 14.03 20 12 20z"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`NameAvatar > NameAvatar name 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 40px;
  height: 40px;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-size: 1.25rem;
  line-height: 1;
  border-radius: 50%;
  overflow: hidden;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #fff;
  background-color: #bdbdbd;
  background-color: #6949a7;
}

.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
  font-size: inherit;
  color: white;
}

<div
    class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault emotion-0"
    name="cosmo"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 emotion-1"
    >
      C
    </p>
  </div>
</DocumentFragment>
`;

exports[`NameAvatar > NameAvatar name dark 1`] = `
<DocumentFragment>
  .emotion-0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 40px;
  height: 40px;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-size: 1.25rem;
  line-height: 1;
  border-radius: 50%;
  overflow: hidden;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #121212;
  background-color: #757575;
  background-color: #6949a7;
}

.emotion-1 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
  font-size: inherit;
  color: black;
}

<div
    class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault emotion-0"
    name="cosmo"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 emotion-1"
    >
      C
    </p>
  </div>
  ,
</DocumentFragment>
`;
